op-node/rollup/derive/blob_data_source.go

```diff
@@ -145,13 +164,14 @@ func (ds *BlobDataSource) open(ctx context.Context) ([]blobOrCalldata, error) {
 // dataAndHashesFromTxs extracts calldata and datahashes from the input transactions and returns them. It
 // creates a placeholder blobOrCalldata element for each returned blob hash that must be populated
 // by fillBlobPointers after blob bodies are retrieved.
-func dataAndHashesFromTxs(txs types.Transactions, config *DataSourceConfig, batcherAddr common.Address, logger log.Logger) ([]blobOrCalldata, []eth.IndexedBlobHash) {
+func dataAndHashesFromTxs(txs types.Transactions, config *DataSourceConfig, batcherAddr common.Address, logger log.Logger, txSucceedMap map[common.Hash]bool) ([]blobOrCalldata, []eth.IndexedBlobHash) {
 	data := []blobOrCalldata{}
 	var hashes []eth.IndexedBlobHash
 	blobIndex := 0 // index of each blob in the block's blob sidecar
 	for _, tx := range txs {
 		// skip any non-batcher transactions or failed transactions
-		if !(isValidBatchTx(tx, config.l1Signer, config.batchInboxAddress, batcherAddr, logger)) {
+		// blobIndex needs to be incremented for both invalid batch tx and failed tx
+		if (!isValidBatchTx(tx, config.l1Signer, config.batchInboxAddress, batcherAddr, logger)) || (txSucceedMap != nil && !txSucceedMap[tx.Hash()]) {
```

Comment:
It is a bit awkward if `txSucceedMap == nil` then all txs are success.

We can use `txFailedMap`, where when it is `nil`, all values will be false.

Reply:
I also considered that, but if some tx is not within the list , `txFailedMap[tx.Hash()]` will also be false, which means successful. It feels more defensive to ensure the tx is within the list and succesful. WDYT ?

Reply:
And there's already a comment about the nil map special case [here](https://github.com/ethstorage/optimism/blob/80f8669ccbc4a4a2eacce8a9b5849681cfe5d12e/op-node/rollup/derive/blob_data_source.go#L77).

Reply:
> I also considered that, but if some tx is not within the list , `txFailedMap[tx.Hash()]` will also be false, which means successful. It feels more defensive to ensure the tx is within the list and succesful. WDYT ?

That should be the bug of the program.  The program should get the correct list, no matter `txFailedMap` or `txSuccessMap`.

Reply:
> Btw there's already a comment about the nil map special case [here](https://github.com/ethstorage/optimism/blob/80f8669ccbc4a4a2eacce8a9b5849681cfe5d12e/op-node/rollup/derive/blob_data_source.go#L77).

The comment should also be added to `dataAndHashesFromTxs` method.

Reply:
> 
> The comment should also be added to `dataAndHashesFromTxs` method.

Done in 3188c12e192be3eb9db2d967b3011db67ef764d8.